File: diffmerge-3/go/dm/merge.go
package dm

import (
	"regexp"
	"strings"
)

// Merge3 performs a three-way merge with conflict detection
func Merge3(base, ours, theirs string, options MergeOptions) MergeResult {
	if options.ConflictStyle == "" {
		options.ConflictStyle = "merge"
	}
	if options.OursLabel == "" {
		options.OursLabel = "ours"
	}
	if options.TheirsLabel == "" {
		options.TheirsLabel = "theirs"
	}
	if options.BaseLabel == "" {
		options.BaseLabel = "base"
	}

	baseLines := SplitLines(base)
	ourLines := SplitLines(ours)
	theirLines := SplitLines(theirs)

	if len(baseLines) == 0 {
		return mergeEmptyBase(ourLines, theirLines, options)
	}

	// Compute what each side did for each base line
	ourDiff := computeDiff3(baseLines, ourLines)
	theirDiff := computeDiff3(baseLines, theirLines)

	result := []string{}
	conflicts := []Conflict{}
	lineNum := 1

	for i := 0; i < len(baseLines); i++ {
		baseLine := baseLines[i]
		ourChange := ourDiff[i]
		theirChange := theirDiff[i]

		// Handle insertions BEFORE this base line (pure additions, not replacements)
		if len(ourChange.insertedBefore) > 0 || len(theirChange.insertedBefore) > 0 {
			insResult, insConf := mergeInsertionsOnly(ourChange.insertedBefore, theirChange.insertedBefore, lineNum, options)
			result = append(result, insResult...)
			if insConf != nil {
				conflicts = append(conflicts, *insConf)
			}
			lineNum += len(insResult)
		}

		// Handle the base line itself
		merged, conf := mergeBaseLine(baseLine, ourChange, theirChange, lineNum, options)
		result = append(result, merged...)
		if conf != nil {
			conflicts = append(conflicts, *conf)
		}
		lineNum += len(merged)
	}

	// Handle insertions at end
	ourChange := ourDiff[len(baseLines)]
	theirChange := theirDiff[len(baseLines)]
	if len(ourChange.insertedBefore) > 0 || len(theirChange.insertedBefore) > 0 {
		insResult, insConf := mergeInsertionsOnly(ourChange.insertedBefore, theirChange.insertedBefore, lineNum, options)
		result = append(result, insResult...)
		if insConf != nil {
			conflicts = append(conflicts, *insConf)
		}
	}

	return MergeResult{
		Content:      strings.Join(result, ""),
		HasConflicts: len(conflicts) > 0,
		Conflicts:    conflicts,
	}
}

// diff3Change represents what happened to a base line
type diff3Change struct {
	kept           bool     // Was this base line kept?
	replacement    []string // If deleted, what replaced it (can be empty for pure deletion)
	insertedBefore []string // Lines inserted BEFORE this position (pure additions)
}

// computeDiff3 computes changes from base to target for each base line
func computeDiff3(baseLines, targetLines []string) map[int]diff3Change {
	changes := make(map[int]diff3Change)

	// Initialize all positions
	for i := 0; i <= len(baseLines); i++ {
		changes[i] = diff3Change{}
	}

	lcs := computeLCS(baseLines, targetLines, false)

	// Build mapping of base lines to target lines (for kept lines)
	keptBase := make(map[int]int) // base index -> target index
	for _, match := range lcs {
		keptBase[match[0]] = match[1]
	}

	// Mark kept lines
	for baseIdx := 0; baseIdx < len(baseLines); baseIdx++ {
		if _, ok := keptBase[baseIdx]; ok {
			c := changes[baseIdx]
			c.kept = true
			changes[baseIdx] = c
		}
	}

	// Now figure out where insertions go
	// Insertions in target are lines not in LCS
	// They need to be associated with either:
	// 1. A deleted base line (as replacements)
	// 2. Before a kept line (as pure insertions)
	// 3. At end

	baseIdx := 0
	targetIdx := 0
	lcsIdx := 0

	for baseIdx < len(baseLines) || targetIdx < len(targetLines) {
		if lcsIdx < len(lcs) {
			nextBaseMatch := lcs[lcsIdx][0]
			nextTargetMatch := lcs[lcsIdx][1]

			// Process gap before this LCS match
			// Collect all target lines before nextTargetMatch
			// Collect all base lines before nextBaseMatch

			gapTargetLines := []string{}
			for targetIdx < nextTargetMatch {
				gapTargetLines = append(gapTargetLines, targetLines[targetIdx])
				targetIdx++
			}

			// Assign gap target lines to deleted base lines or as insertions before kept line
			gapBaseStart := baseIdx
			gapBaseEnd := nextBaseMatch

			if gapBaseEnd > gapBaseStart {
				// There are deleted base lines - associate target lines as replacements
				// Assign all to the first deleted base line
				for bi := gapBaseStart; bi < gapBaseEnd; bi++ {
					if bi == gapBaseStart {
						c := changes[bi]
						c.replacement = gapTargetLines
						changes[bi] = c
					}
					baseIdx++
				}
			} else {
				// No deleted base lines - these are pure insertions before the kept line
				c := changes[nextBaseMatch]
				c.insertedBefore = gapTargetLines
				changes[nextBaseMatch] = c
			}

			// Skip past the matched lines
			baseIdx++
			targetIdx++
			lcsIdx++
		} else {
			// No more LCS matches - remaining target lines go to end position
			// Remaining base lines are deleted
			gapTargetLines := []string{}
			for targetIdx < len(targetLines) {
				gapTargetLines = append(gapTargetLines, targetLines[targetIdx])
				targetIdx++
			}

			gapBaseStart := baseIdx
			gapBaseEnd := len(baseLines)

			if gapBaseEnd > gapBaseStart {
				// Deleted base lines at end - associate target lines as replacements
				for bi := gapBaseStart; bi < gapBaseEnd; bi++ {
					if bi == gapBaseStart {
						c := changes[bi]
						c.replacement = gapTargetLines
						changes[bi] = c
					}
					baseIdx++
				}
			} else {
				// Pure insertions at end
				c := changes[len(baseLines)]
				c.insertedBefore = gapTargetLines
				changes[len(baseLines)] = c
			}
			break
		}
	}

	return changes
}

// mergeBaseLine merges changes to a specific base line
func mergeBaseLine(baseLine string, ourChange, theirChange diff3Change, lineNum int, options MergeOptions) ([]string, *Conflict) {
	// Both kept
	if ourChange.kept && theirChange.kept {
		return []string{baseLine}, nil
	}

	// One kept, one deleted (with or without replacement)
	if ourChange.kept && !theirChange.kept {
		if len(theirChange.replacement) == 0 {
			// Theirs just deleted - accept deletion
			return nil, nil
		} else {
			// Theirs modified - take their modification
			return theirChange.replacement, nil
		}
	}

	if theirChange.kept && !ourChange.kept {
		if len(ourChange.replacement) == 0 {
			// Ours just deleted - accept deletion
			return nil, nil
		} else {
			// Ours modified - take our modification
			return ourChange.replacement, nil
		}
	}

	// Both deleted
	ourRep := strings.Join(ourChange.replacement, "")
	theirRep := strings.Join(theirChange.replacement, "")

	if ourRep == theirRep {
		// Same modification (or both just deleted)
		return ourChange.replacement, nil
	}

	// Different modifications - CONFLICT
	conflict := Conflict{
		Base:      baseLine,
		Ours:      ourRep,
		Theirs:    theirRep,
		StartLine: lineNum,
		EndLine:   lineNum,
	}
	return formatConflict(conflict, options), &conflict
}

func mergeInsertionsOnly(ourIns, theirIns []string, lineNum int, options MergeOptions) ([]string, *Conflict) {
	ourContent := strings.Join(ourIns, "")
	theirContent := strings.Join(theirIns, "")

	if ourContent == "" && theirContent == "" {
		return nil, nil
	}
	if ourContent == "" {
		return theirIns, nil
	}
	if theirContent == "" {
		return ourIns, nil
	}
	if ourContent == theirContent {
		return ourIns, nil
	}

	conflict := Conflict{
		Base:      "",
		Ours:      ourContent,
		Theirs:    theirContent,
		StartLine: lineNum,
		EndLine:   lineNum,
	}
	return formatConflict(conflict, options), &conflict
}

func mergeEmptyBase(ourLines, theirLines []string, options MergeOptions) MergeResult {
	if len(ourLines) == 0 && len(theirLines) == 0 {
		return MergeResult{Content: "", HasConflicts: false, Conflicts: []Conflict{}}
	}
	if len(ourLines) == 0 {
		return MergeResult{Content: strings.Join(theirLines, ""), HasConflicts: false, Conflicts: []Conflict{}}
	}
	if len(theirLines) == 0 {
		return MergeResult{Content: strings.Join(ourLines, ""), HasConflicts: false, Conflicts: []Conflict{}}
	}

	ourContent := strings.Join(ourLines, "")
	theirContent := strings.Join(theirLines, "")

	if ourContent == theirContent {
		return MergeResult{Content: ourContent, HasConflicts: false, Conflicts: []Conflict{}}
	}

	conflict := Conflict{Base: "", Ours: ourContent, Theirs: theirContent, StartLine: 1, EndLine: 1}
	result := formatConflict(conflict, options)
	return MergeResult{Content: strings.Join(result, ""), HasConflicts: true, Conflicts: []Conflict{conflict}}
}

func formatConflict(conflict Conflict, options MergeOptions) []string {
	result := []string{}

	result = append(result, "<<<<<<< "+options.OursLabel+"\n")
	if conflict.Ours != "" {
		result = append(result, SplitLines(conflict.Ours)...)
	}

	if options.ConflictStyle == "diff3" {
		result = append(result, "||||||| "+options.BaseLabel+"\n")
		if conflict.Base != "" {
			result = append(result, SplitLines(conflict.Base)...)
		}
	}

	result = append(result, "=======\n")
	if conflict.Theirs != "" {
		result = append(result, SplitLines(conflict.Theirs)...)
	}
	result = append(result, ">>>>>>> "+options.TheirsLabel+"\n")

	return result
}

func HasConflicts(content string) bool {
	return strings.Contains(content, "<<<<<<<") &&
		strings.Contains(content, "=======") &&
		strings.Contains(content, ">>>>>>>")
}

func ExtractConflicts(content string) []Conflict {
	conflicts := []Conflict{}
	lines := strings.Split(content, "\n")

	i := 0
	lineNum := 1
	for i < len(lines) {
		if strings.HasPrefix(lines[i], "<<<<<<<") {
			conflict, endI := parseConflict(lines, i)
			conflict.StartLine = lineNum
			conflict.EndLine = lineNum + (endI - i)
			conflicts = append(conflicts, conflict)
			lineNum += endI - i + 1
			i = endI + 1
		} else {
			lineNum++
			i++
		}
	}

	return conflicts
}

func parseConflict(lines []string, startI int) (Conflict, int) {
	conflict := Conflict{}
	i := startI

	if i < len(lines) && strings.HasPrefix(lines[i], "<<<<<<<") {
		i++
	}

	oursContent := []string{}
	for i < len(lines) && !strings.HasPrefix(lines[i], "|||||||") && !strings.HasPrefix(lines[i], "=======") {
		oursContent = append(oursContent, lines[i])
		i++
	}
	conflict.Ours = strings.Join(oursContent, "\n")
	if len(oursContent) > 0 {
		conflict.Ours += "\n"
	}

	if i < len(lines) && strings.HasPrefix(lines[i], "|||||||") {
		i++
		baseContent := []string{}
		for i < len(lines) && !strings.HasPrefix(lines[i], "=======") {
			baseContent = append(baseContent, lines[i])
			i++
		}
		conflict.Base = strings.Join(baseContent, "\n")
		if len(baseContent) > 0 {
			conflict.Base += "\n"
		}
	}

	if i < len(lines) && strings.HasPrefix(lines[i], "=======") {
		i++
	}

	theirsContent := []string{}
	for i < len(lines) && !strings.HasPrefix(lines[i], ">>>>>>>") {
		theirsContent = append(theirsContent, lines[i])
		i++
	}
	conflict.Theirs = strings.Join(theirsContent, "\n")
	if len(theirsContent) > 0 {
		conflict.Theirs += "\n"
	}

	return conflict, i
}

func ResolveConflict(content string, index int, resolution string) string {
	conflicts := ExtractConflicts(content)
	if index < 0 || index >= len(conflicts) {
		return content
	}

	conflict := conflicts[index]
	lines := strings.Split(content, "\n")

	conflictCount := 0
	startLine := -1
	endLine := -1

	for i := 0; i < len(lines); i++ {
		if strings.HasPrefix(lines[i], "<<<<<<<") {
			if conflictCount == index {
				startLine = i
			}
			conflictCount++
		}
		if strings.HasPrefix(lines[i], ">>>>>>>") && startLine >= 0 && endLine < 0 {
			endLine = i
			break
		}
	}

	if startLine < 0 || endLine < 0 {
		return content
	}

	var resolutionContent string
	switch resolution {
	case "ours":
		resolutionContent = strings.TrimSuffix(conflict.Ours, "\n")
	case "theirs":
		resolutionContent = strings.TrimSuffix(conflict.Theirs, "\n")
	case "base":
		resolutionContent = strings.TrimSuffix(conflict.Base, "\n")
	default:
		resolutionContent = strings.TrimSuffix(resolution, "\n")
	}

	resultLines := []string{}
	resultLines = append(resultLines, lines[:startLine]...)
	if resolutionContent != "" {
		resultLines = append(resultLines, strings.Split(resolutionContent, "\n")...)
	}
	resultLines = append(resultLines, lines[endLine+1:]...)

	return strings.Join(resultLines, "\n")
}

var conflictMarkerPattern = regexp.MustCompile(`^<<<<<<<|^=======|^>>>>>>>`)

File: diffmerge-2/py/diffmerge/merge.py
"""Three-way merge functionality for diffmerge library."""

import re
from typing import List, Optional, Tuple
from .types import MergeResult, MergeOptions, Conflict
from .utils import split_lines


def _lcs_indices(a: List[str], b: List[str]) -> List[Tuple[int, int]]:
    """Compute LCS and return list of (index_in_a, index_in_b) pairs."""
    m, n = len(a), len(b)
    if m == 0 or n == 0:
        return []

    # Build LCS length table
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    # Backtrack
    result = []
    i, j = m, n
    while i > 0 and j > 0:
        if a[i - 1] == b[j - 1]:
            result.append((i - 1, j - 1))
            i -= 1
            j -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1
    result.reverse()
    return result


def merge3(
    base: str,
    ours: str,
    theirs: str,
    options: Optional[MergeOptions] = None
) -> MergeResult:
    """Three-way merge with conflict detection."""
    opts = options or {}
    conflict_style = opts.get("conflict_style", "merge")
    ours_label = opts.get("ours_label", "ours")
    theirs_label = opts.get("theirs_label", "theirs")
    base_label = opts.get("base_label", "base")

    base_lines = split_lines(base)
    ours_lines = split_lines(ours)
    theirs_lines = split_lines(theirs)

    # Find LCS of base with ours and theirs
    base_ours_lcs = _lcs_indices(base_lines, ours_lines)
    base_theirs_lcs = _lcs_indices(base_lines, theirs_lines)

    # Build maps: base_idx -> ours_idx, base_idx -> theirs_idx
    base_to_ours = {b: o for b, o in base_ours_lcs}
    base_to_theirs = {b: t for b, t in base_theirs_lcs}

    result_lines: List[str] = []
    conflicts: List[Conflict] = []

    base_idx = 0
    ours_idx = 0
    theirs_idx = 0

    while base_idx < len(base_lines) or ours_idx < len(ours_lines) or theirs_idx < len(theirs_lines):
        # Find next sync point in base that exists in both ours and theirs
        sync_base = None
        for bi in range(base_idx, len(base_lines)):
            if bi in base_to_ours and bi in base_to_theirs:
                sync_base = bi
                break

        if sync_base is not None:
            sync_ours = base_to_ours[sync_base]
            sync_theirs = base_to_theirs[sync_base]

            # Collect changes before sync point
            ours_before = ours_lines[ours_idx:sync_ours]
            theirs_before = theirs_lines[theirs_idx:sync_theirs]
            base_before = base_lines[base_idx:sync_base]

            # Handle the region before sync
            merged, region_conflicts = _merge_region(
                base_before, ours_before, theirs_before,
                conflict_style, ours_label, theirs_label, base_label,
                len(result_lines) + 1
            )
            result_lines.extend(merged)
            conflicts.extend(region_conflicts)

            # Add the sync line
            result_lines.append(base_lines[sync_base])
            base_idx = sync_base + 1
            ours_idx = sync_ours + 1
            theirs_idx = sync_theirs + 1
        else:
            # No more sync points - handle remaining lines
            ours_remaining = ours_lines[ours_idx:]
            theirs_remaining = theirs_lines[theirs_idx:]
            base_remaining = base_lines[base_idx:]

            merged, region_conflicts = _merge_region(
                base_remaining, ours_remaining, theirs_remaining,
                conflict_style, ours_label, theirs_label, base_label,
                len(result_lines) + 1
            )
            result_lines.extend(merged)
            conflicts.extend(region_conflicts)
            break

    content = "".join(result_lines)
    return {
        "content": content,
        "has_conflicts": len(conflicts) > 0,
        "conflicts": conflicts
    }


def _merge_region(
    base: List[str],
    ours: List[str],
    theirs: List[str],
    conflict_style: str,
    ours_label: str,
    theirs_label: str,
    base_label: str,
    start_line: int
) -> Tuple[List[str], List[Conflict]]:
    """Merge a region between sync points."""
    result: List[str] = []
    conflicts: List[Conflict] = []

    # Simple cases
    if not base and not ours and not theirs:
        return result, conflicts

    if ours == theirs:
        # Both same - use either
        return list(ours), conflicts

    if ours == base:
        # Only theirs changed
        return list(theirs), conflicts

    if theirs == base:
        # Only ours changed
        return list(ours), conflicts

    # When all three have the same length, try line-by-line merge
    if len(base) == len(ours) == len(theirs):
        can_merge = True
        merged_lines: List[str] = []
        for i in range(len(base)):
            b, o, t = base[i], ours[i], theirs[i]
            if o == t:
                # Both same (could be same as base or both changed same)
                merged_lines.append(o)
            elif o == b:
                # Only theirs changed
                merged_lines.append(t)
            elif t == b:
                # Only ours changed
                merged_lines.append(o)
            else:
                # Both changed differently - conflict
                can_merge = False
                break

        if can_merge:
            return merged_lines, conflicts

    # Fall back to treating entire region as conflict
    base_content = "".join(base)
    ours_content = "".join(ours)
    theirs_content = "".join(theirs)

    conflict_start = start_line
    result.append(f"<<<<<<< {ours_label}\n")
    result.extend(ours)
    if conflict_style == "diff3":
        result.append(f"||||||| {base_label}\n")
        result.extend(base)
    result.append("=======\n")
    result.extend(theirs)
    result.append(f">>>>>>> {theirs_label}\n")
    conflict_end = start_line + len(result) - 1

    conflicts.append({
        "base": base_content,
        "ours": ours_content,
        "theirs": theirs_content,
        "start_line": conflict_start,
        "end_line": conflict_end
    })

    return result, conflicts


def has_conflicts(content: str) -> bool:
    """Check if content contains conflict markers."""
    has_start = "<<<<<<" in content
    has_middle = "======" in content
    has_end = ">>>>>>" in content
    return has_start and has_middle and has_end


def extract_conflicts(content: str) -> List[Conflict]:
    """Extract conflict regions from merged content."""
    conflicts: List[Conflict] = []
    lines = content.split("\n")

    i = 0
    while i < len(lines):
        line = lines[i]

        if line.startswith("<<<<<<<"):
            start_line = i + 1
            ours_lines: List[str] = []
            base_lines: List[str] = []
            theirs_lines: List[str] = []

            i += 1
            section = "ours"

            while i < len(lines):
                line = lines[i]
                if line.startswith("|||||||"):
                    section = "base"
                    i += 1
                    continue
                elif line.startswith("======="):
                    section = "theirs"
                    i += 1
                    continue
                elif line.startswith(">>>>>>>"):
                    end_line = i + 1
                    conflicts.append({
                        "ours": "\n".join(ours_lines) + ("\n" if ours_lines else ""),
                        "base": "\n".join(base_lines) + ("\n" if base_lines else ""),
                        "theirs": "\n".join(theirs_lines) + ("\n" if theirs_lines else ""),
                        "start_line": start_line,
                        "end_line": end_line
                    })
                    break
                else:
                    if section == "ours":
                        ours_lines.append(line)
                    elif section == "base":
                        base_lines.append(line)
                    else:
                        theirs_lines.append(line)
                i += 1
        i += 1

    return conflicts


def resolve_conflict(
    content: str,
    conflict_index: int,
    resolution: str
) -> str:
    """Resolve a specific conflict in the content."""
    conflicts = extract_conflicts(content)

    if conflict_index < 0 or conflict_index >= len(conflicts):
        return content

    conflict = conflicts[conflict_index]

    if resolution == "ours":
        replacement = conflict["ours"]
    elif resolution == "theirs":
        replacement = conflict["theirs"]
    elif resolution == "base":
        replacement = conflict["base"]
    else:
        replacement = resolution

    lines = content.split("\n")
    result_lines: List[str] = []
    current_conflict = 0
    i = 0

    while i < len(lines):
        line = lines[i]

        if line.startswith("<<<<<<<"):
            if current_conflict == conflict_index:
                while i < len(lines) and not lines[i].startswith(">>>>>>>"):
                    i += 1
                i += 1

                if replacement:
                    rep_lines = replacement.rstrip("\n").split("\n")
                    for rline in rep_lines:
                        result_lines.append(rline)
                current_conflict += 1
            else:
                result_lines.append(line)
                i += 1
                current_conflict += 1
        else:
            result_lines.append(line)
            i += 1

    return "\n".join(result_lines)
